src/app/components/problems/problem-set/problem-set.component.ts

```diff
@@ -173,4 +201,63 @@ export class ProblemSetComponent implements OnInit {
         this.deleteQuestionId = questionId;
         this.modalService.open(content, {ariaLabelledBy: 'modal-basic-title', centered: true});
     }
+
+    private generateJSONURI(arg: {
+        filename: string,
+        text: string
+    }) {
+        this.filename = arg.filename;
+        this.jsonUri = this.sanitizer.bypassSecurityTrustUrl("data:text/json;charset=UTF-8," + encodeURIComponent(arg.text));
+    }
+
+    getQuestionJson(question: Question): SafeUrl {
+        return this.sanitizer.bypassSecurityTrustUrl("data:text/json;charset=UTF-8," + encodeURIComponent(JSON.stringify(question)));
+    }
+
+    getFileName(id: number): string {
+        const timestamp = formatDate(new Date(), 'yyyy/MM/dd_HH:mm:ss', 'en');
+        return `question-${id}-${timestamp}.json`;
+    }
+
+    downloadQuestion(id: number) : SafeUrl {
+        return this.getQuestionJson(this.retrievedQuestions?.filter(c => c.id === id)[0]);
+    }
+
+    onFileChanged(target: EventTarget): void {
+        const selectedFile = (<HTMLInputElement>target).files[0];
+        const fileReader = new FileReader();
+        fileReader.readAsText(selectedFile, "UTF-8");
+        fileReader.onload = () => {
+            if (typeof fileReader.result === "string") {
+                this.parsedQuestions = JSON.parse(fileReader.result);
+            }
+        };
+        fileReader.onerror = (error) => {
+            console.log(error);
+            this.toastr.error('Error occurred while reading file');
+        };
+    }
+
+    uploadQuestions(): void {
+        const questions = this.parsedQuestions;
+        if (questions == null)
+            return;
+        if (typeof questions[Symbol.iterator] === 'function') {
+            for (const question of questions) {
+                this.uploadQuestion(question);
+            }
+        }
+        else
+            this.uploadQuestion(questions as unknown as Question);
```

Comment:
Why do we need two `as`?

Reply:
![image](https://user-images.githubusercontent.com/16740461/121820722-380d3280-cc49-11eb-997f-d1e8974c127d.png)

Reply:
Why do you have to convert `Question[]` to `Question`? That doesn't make sense, does it?

Reply:
It's converted to just `Question` to avoid storing the downloaded single question as an array with length 1. A single question can't be iterated over, so we convert it here so that it will work with the upload function. I tried `questions[0]` but that gets the first property of the question object not the entire object, so only when there is a single question in the file it will convert from `Question[]` to `Question`